src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -166,6 +179,221 @@ instance (NFData t, NFData r) => NFData (SnapMergingRunState t r) where
+
+instance NFData r => NFData (SnapPreExistingRun r) where
+  rnf (SnapPreExistingRun a) = rnf a
+  rnf (SnapPreExistingMergingRun a b c d) =
+      rnf a `seq` rnf b `seq` rnf c `seq` rnf d
+
+{-------------------------------------------------------------------------------
+  Opening from merging tree snapshot format
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE fromSnapMergingTree ::
+     ActionRegistry IO
+  -> HasFS IO h
+  -> HasBlockIO IO h
+  -> TableConfig
+  -> UniqCounter IO
+  -> ResolveSerialisedValue
+  -> ActiveDir
+  -> SnapMergingTree (Ref (Run IO h))
+  -> IO (Ref (MT.MergingTree IO h))
+  #-}
+-- | Duplicates runs and re-creates merging runs.
+fromSnapMergingTree ::
+     forall m h. (MonadMask m, MonadMVar m, MonadSTM m, MonadST m)
+  => ActionRegistry m
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> TableConfig
+  -> UniqCounter m
+  -> ResolveSerialisedValue
+  -> ActiveDir
+  -> SnapMergingTree (Ref (Run m h))
+  -> m (Ref (MT.MergingTree m h))
+fromSnapMergingTree reg hfs hbio conf@TableConfig{..} uc resolve dir (SnapMergingTree snapTreeState) =
+    fromSnapTreeState snapTreeState
+  where
+    mkPath = RunFsPaths (getActiveDir dir)
+
+    dupRun :: RefCounted m obj => Ref obj -> m (Ref obj)
+    dupRun r = withRollback reg (dupRef r) releaseRef
+
+    -- Used for inductive calls
+    recurrance :: SnapMergingTree (Ref (Run m h)) -> m (Ref (MT.MergingTree m h))
+    recurrance = fromSnapMergingTree reg hfs hbio conf uc resolve dir
+
+    fromSnapTreeState :: SnapMergingTreeState (Ref (Run m h)) -> m (Ref (MT.MergingTree m h))
+    fromSnapTreeState (SnapCompletedTreeMerge run) = dupRun run >>= \run' ->
+      MT.newPendingLevelMerge [MT.PreExistingRun run'] Nothing
+    fromSnapTreeState (SnapPendingTreeMerge pMerge) = case pMerge of
+      SnapPendingLevelMerge peRuns maybeMergeTree -> do
+        peRuns' <- traverse fromSnapPreExistingRun peRuns
+        maybeMergeTree' <- traverse recurrance maybeMergeTree
+        MT.newPendingLevelMerge peRuns' maybeMergeTree'
+      SnapPendingUnionMerge mergeTrees -> MT.newPendingUnionMerge =<< traverse recurrance mergeTrees
+    fromSnapTreeState (SnapOngoingTreeMerge nr ne (SuppliedCredits sc) smrs) = do
+        ongoingState <- getSnapMergingRunState nr ne sc smrs
+        mergeState <- newMVar $! MT.OngoingTreeMerge ongoingState
+        newRef (MT.finalise mergeState) $ \mergeRefCounter ->
+          MT.MergingTree { mergeState, mergeRefCounter }
+
+    getSnapMergingRunState
+      :: forall t.
+         MR.IsMergeType t
+      => NumRuns
+      -> NumEntries
+      -> Int
+      -> SnapMergingRunState t (Ref (Run m h))
+      -> m (Ref (MR.MergingRun t m h))
+    getSnapMergingRunState nr ne sc smrs = do
+          let ln = LevelNo 0
+              caching = diskCachePolicyForLevel confDiskCachePolicy ln
+              alloc = bloomFilterAllocForLevel conf ln
+              indexType = indexTypeForRun confFencePointerIndex
```

Comment:
I've seen this code before. We should somehow be able to share it.

I think it lives in IncomingRun, which also wraps a merging run.